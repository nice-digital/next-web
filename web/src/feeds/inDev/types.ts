import { ProductGroup, ProductTypeAcronym } from "@/feeds/publications/types";

export enum FeedPath {
	AllProjects = "/gidprojects/all",
	InConsultationProjects = "/inconsultationprojects",
}

export enum ProjectStatus {
	"Discontinued" = "Discontinued",
	"Suspended" = "Suspended",
	"Referred" = "Referred",
	"InProgress" = "InProgress",
	"Proposed" = "Proposed",
	"ImpactedByCOVID19" = "ImpactedByCOVID19",
}

/** Not a definitive list - this may change over time */
export enum ProjectProcess {
	"MTA" = "MTA",
	"IP" = "IP",
	"TAG" = "TAG",
	"MT" = "MT",
	"DT" = "DT",
	"STAPre2018" = "STA pre-2018",
	"STA2018" = "STA 2018",
	"QSD" = "QSD",
	"CG" = "CG",
	"APG" = "APG",
	"SG" = "SG",
	"HST" = "HST",
	"MIB" = "MIB",
	"STAReview" = "STA Review",
	"PHG" = "PHG",
	"MTAReview" = "MTA Review",
	"CDFReview" = "CDF Review",
	"NG" = "NG",
	"APA" = "APA",
	"ES" = "ES",
	"SC" = "SC",
	"COVID19RapidGuideline" = "COVID-19 rapid guideline",
}

export type ProjectGroup = ProductGroup.Guidance | ProductGroup.Advice;

export type GIDReference = `GID-${Uppercase<string>}${number}`;

export interface AllProjects {
	_links: {
		// In theory InDev supports pagination, but in reality the gidprojects/all feed always returns all projects and a page size of 2048
		first: [
			{
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
		last: [
			{
				// This looks like a bug in InDev as Page=0
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
		self: [
			{
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
	};
	_embedded: {
		"nice.indev:indevelopment-project": Project[];
	};
}

export interface Project {
	_links: {
		self: [
			{
				href: `/project/${GIDReference}`;
			}
		];
	};
	/** e.g. 01000000-0000-001B-0000-000000005402 */
	ETag: string;
	/** Project reference e.g. GID-TAG377 */
	Reference: GIDReference;
	/** Acronym (usually 2 or 3 letters) for the type of the project e.g "TA", "IPG" etc */
	ProjectType: ProductTypeAcronym;
	/** The name of the project type e.g. "Technology appraisal guidance" for TA or "Interventional procedures guidance" for "IPG" */
	ProductTypeName: string;
	Process: ProjectProcess | string;
	/** Project title e.g. "Atrial fibrillation - ximelagatran [ID376]" or "Clinically isolated syndrome - beta interferons and glatiramer acetate [ID109]" */
	Title: string;
	Status: ProjectStatus;
	ProjectGroup: ProjectGroup;
	/**
	 * `null` for unpublished projects or an ISO formatted date string like `2020-12-15T00:00:00`
	 */
	PublishedDate: string | null;
	/**
	 * An ISO formatted date string like `2020-09-22T13:51:48.8447645`
	 */
	LastModifiedDate: string;
	/**
	 * Null for unpublished projects or an ISO formatted date string like `2020-10-05T12:27:21.5437767`
	 */
	FirstGoLiveDate: string | null;
	/** The date the project was created in in dev */
	CreatedDate: string;
	DevelopedAs: string | null;
	/** Almost always the same as `DevelopedAs` */
	RelevantTo: string | null;
	IdNumber: `${number}` | null;
	/** E.g. "BMJ Evidence Centre, BMJ Group", "ScHARR" or "Warwick Evidence" etc */
	EvidenceAssessmentGroup: string | null;
	/** TODO Always seems to be an empty array */
	AreasOfInterestList: [];
}

export interface InConsultationProjects {
	_links: {
		self: [
			{
				href: "/inconsultationprojects?Page=0&PageSize=2048";
			}
		];
	};
	// Sometimes there are no consultations running, in which case `_embedded` is not in the feed at all
	_embedded?: {
		"nice.indev:inconsultation-product": Consultation[];
	};
	ETag: null;
	// Page and size is hard coded here
	Page: 0;
	PageSize: 2048;
	TotalItems: number;
}

export interface Consultation {
	_links: {
		self: [
			{
				href: `/guidance/${GIDReference}/consultation/html-content${
					| number
					| ""}`;
			}
		];
	};
	ETag: null;
	Reference: GIDReference;
	Title: string;
	ConsultationName: string;
	/** ISO formatted date like `2021-06-18T00:00:00` */
	StartDate: string;
	/** ISO formatted date like `2021-06-18T00:00:00` */
	EndDate: string;
	ConsultationType: string;
	ResourceTitleId: string;
	ProjectType: ProjectProcess | string;
	ProductTypeName: string;
	ShowExpressionOfInterestSubmissionQuestion: boolean;
	DevelopedAs: null;
	RelevantTo: null;
	ConsultationId: number;
	/** E.g. QSD, IP etc */
	Process: string;
	HasDocumentsWhichAllowConsultationComments: boolean;
	IsCHTE: boolean;
	/** E.g. HSCTeam or CHTETeam */
	AllowedRole: string;
	FirstConvertedDocumentId: null;
	FirstChapterSlugOfFirstConvertedDocument: null;
	PartiallyUpdatedProjectReference: null;
	OrigProjectReference: null;
	AreasOfInterestList: [];
}
