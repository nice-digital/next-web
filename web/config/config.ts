import config from "config";

export interface SettingsConfig {
	/** Name of the environment e.g. "dev" or "test" */
	environment: string;
}

export const settings = config.get<SettingsConfig>("settings");
