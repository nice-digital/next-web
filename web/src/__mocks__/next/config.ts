import nextConfig from "./../../../next.config";

const getConfig = (): unknown => ({
	publicRuntimeConfig: nextConfig.publicRuntimeConfig,
	serverRuntimeConfig: nextConfig.serverRuntimeConfig,
});

export default getConfig;
