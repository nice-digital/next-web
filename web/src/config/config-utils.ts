export const getNextPublicEnvVars = async (): Promise<{
	NEXT_PUBLIC_BASE_URL: string;
	NEXT_PUBLIC_AUTH_ENVIRONMENT: "test" | "live" | "beta" | "local";
	SANTA_BUILD_NUMBER: string;
	NEXT_PUBLIC_BUILD_NUMBER: string;
	NEXT_PUBLIC_SEARCH_BASE_URL: string;
	PUBLIC_DENY_ROBOTS: string;
	NEXT_PUBLIC_JOHN: string;
	PUBLIC_JOHN: string;
	PUBLIC_NEW_JOHN: string;
	SANTA_BOTH: string;
	SANTA_TC: string;
	SANTA_ENV: string;
	SANTA_NONE: string;
	NEXT_PUBLIC_SANTA_BOTH: string;
	NEXT_PUBLIC_SANTA_TC: string;
	NEXT_PUBLIC_SANTA_ENV: string;
	NEXT_PUBLIC_SANTA_NONE: string;
}> => {
	if (typeof window === "undefined") {
		// Server-side: Use environment variables directly
		return {
			NEXT_PUBLIC_BASE_URL: process.env.NEXT_PUBLIC_BASE_URL || "",
			NEXT_PUBLIC_AUTH_ENVIRONMENT:
				(process.env.NEXT_PUBLIC_AUTH_ENVIRONMENT as
					| "test"
					| "live"
					| "beta"
					| "local") || "local",
			SANTA_BUILD_NUMBER:
				process.env.SANTA_BUILD_NUMBER || "config-utils-fallback",
			NEXT_PUBLIC_BUILD_NUMBER: process.env.NEXT_PUBLIC_BUILD_NUMBER || "0",
			NEXT_PUBLIC_SEARCH_BASE_URL:
				process.env.NEXT_PUBLIC_SEARCH_BASE_URL || "",
			PUBLIC_DENY_ROBOTS: process.env.PUBLIC_DENY_ROBOTS || "",
			NEXT_PUBLIC_JOHN: process.env.NEXT_PUBLIC_JOHN || "",
			PUBLIC_JOHN: process.env.PUBLIC_JOHN || "",
			PUBLIC_NEW_JOHN: process.env.PUBLIC_NEW_JOHN || "",
			SANTA_BOTH: process.env.SANTA_BOTH || "",
			SANTA_TC: process.env.SANTA_TC || "",
			SANTA_ENV: process.env.SANTA_ENV || "",
			SANTA_NONE: process.env.SANTA_NONE || "",
			NEXT_PUBLIC_SANTA_BOTH: process.env.NEXT_PUBLIC_SANTA_BOTH || "",
			NEXT_PUBLIC_SANTA_TC: process.env.NEXT_PUBLIC_SANTA_TC || "",
			NEXT_PUBLIC_SANTA_ENV: process.env.NEXT_PUBLIC_SANTA_ENV || "",
			NEXT_PUBLIC_SANTA_NONE: process.env.NEXT_PUBLIC_SANTA_NONE || "",
		};
	}

	// Client-side: Fetch from API
	const response = await fetch(
		`${process.env.NEXT_PUBLIC_BASE_URL || ""}/api/env`,
		{
			cache: "no-store",
		}
	);

	let env;
	try {
		env = await response.json();
	} catch (error) {
		console.error("Invalid JSON response:", error);
		throw new Error("Failed to fetch environment variables");
	}

	return env;
};
