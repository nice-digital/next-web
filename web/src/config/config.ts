//import config from "config";
import getConfig from "next/config";

import type { InitialiseOptions as SearchClientInitOptions } from "@nice-digital/search-client";

const { publicRuntimeConfig, serverRuntimeConfig } = getConfig();

export interface SearchConfig {
	/** The base URL of the Single Search Endpoint (SSE) e.g. https://beta-search-api.nice.org.uk/api/ */
	readonly baseURL: SearchClientInitOptions["baseURL"];
}

export interface CacheControlConfig {
	/** The default value for the cache-control header */
	readonly defaultCacheHeader: string;
}

export interface JotFormPublicConfig {
	/** The full base URL for NICE's JotForm instance, including protocol but excluding trailing slash */
	readonly baseURL: string;
}

/**
 * Public run time config, available to both client and server
 */
export interface PublicConfig {
	/** The version number of the current build (or Octopus.Release.Number in Octo) */
	readonly buildNumber: string;

	/** Name of the environment e.g. _dev_ or _test_ */
	readonly environment: string;

	/** The base URL of the website including protocol and port e.g. http://localhost:3000 for local dev or http://dev.nice.org.uk.
	 *
	 * **Note** the lack of trailing slash! It will get prepended to paths that start with a slash. */
	readonly baseURL: string;

	/**
	 * The absolute URL to the NICE cookie banner script include
	 */
	readonly cookieBannerScriptURL: string;

	/**
	 * The base URL to the deployed NextJS _public_ folder, see https://nextjs.org/docs/basic-features/static-file-serving.
	 * Empty string means relative to the deployed app. Set it to an absolute path in _config.yml_ to use a CDN.
	 */
	readonly publicBaseURL: string;

	readonly search: SearchConfig;

	/** Config for cache-control response headers */
	readonly cacheControl: CacheControlConfig;

	/** Public config for JotForm. Note, there's also a server config object with the secret API key */
	readonly jotForm: JotFormPublicConfig;

	/** Public config for Storyblok */
	readonly storyblok: StoryblokConfig;
}

/**
 * Config for object cache
 */
export interface CacheConfig {
	/** The prefix for cache keys */
	readonly keyPrefix: string;

	/** Path to the folder used for file system disk cache */
	readonly filePath: string;

	/** The default TTL (time to live), in seconds, of cache entries */
	readonly defaultTTL: number;

	/** The longer TTL (time to live), in seconds, for long-lived cache entries that rarely change */
	readonly longTTL: number;

	/** Threshold for TTL, in seconds, below which object caches are refreshed in the background as per https://edibleco.de/3hyfpGG */
	readonly refreshThreshold: number;
}

export interface FeedConfig {
	/** The origin URL of the feeds */
	readonly origin: string;

	/** The API key for accessing the feed */
	readonly apiKey: string;
}

export interface FeedsConfig {
	/** Feed config for publications */
	readonly publications: FeedConfig;

	/** Feed config for indev */
	readonly inDev: FeedConfig;

	/** Feed config for JotForm */
	readonly jotForm: Pick<FeedConfig, "apiKey">;
}

export interface StoryblokConfig {
	readonly accessToken: string;
	readonly ocelotEndpoint: string;
	readonly enableRootCatchAll: string;
}

/**
 * Server-only run time config, useful for secrets etc
 */
export interface ServerConfig {
	cache: CacheConfig;
	feeds: FeedsConfig;
	storyblok: StoryblokConfig;
}

export { publicRuntimeConfig, serverRuntimeConfig };
