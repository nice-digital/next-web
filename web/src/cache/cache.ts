import { caching, Cache } from "cache-manager";
import store from "cache-manager-fs-hash";

import { serverRuntimeConfig } from "@/config";

const { keyPrefix, defaultTTL, refreshThreshold, filePath } =
	serverRuntimeConfig.cache;

/**
 * Returns a cache key generated from the given prefix and postfix.
 *
 * Note: it's also prefixed with a global cache key prefix that comes from config.
 *
 * @example
 * 	const cacheKey = getCacheKey("publications", "products")
 * 	// returns "next-web:local:publications:products";
 *
 * @param groupKey The prefix for a group of cached objects e.g. "publications"
 * @param itemKey The individual cache key for the entry
 * @returns The generated cache key, prefixed with the global cache key prefix.
 */
export const getCacheKey = (groupKey: string, itemKey: string): string =>
	`${keyPrefix}:${groupKey}:${itemKey}`;

/**
 * A file-system/disk cache using cache-manager-fs-hash.
 *
 * We use fs-hash because it supports clustering, via lock files. We use clustering
 * via PM2 in production, so a memory cache wouldn't be suitable: memory is isolated to each
 * process in cluster mode.
 *
 * FS is slower than in-memory but the tradeoff is not having cache syncing issues and persisting cache between
 * app restarts/deployments. The cache will hopefully be replaced by a separate API gateway microservice
 * ("backend for front end") layer, backed by Redis. But FS will do for the time-being.
 */
export const cache: Readonly<Cache> = caching({
	store,
	options: {
		path: filePath,
		ttl: defaultTTL,
		subdirs: true,
	},
	// Using a sensible refreshThreshold allows use to almost follow a SWR (stale while relvalidate) approach.
	// 'Almost' meaning we don't relvalidate on every cache access as we don't need to be that up-to-date but we
	// revalidate regularly. We have enough traffic that this should mean a lot of the time things are always in cache.
	refreshThreshold,
});
