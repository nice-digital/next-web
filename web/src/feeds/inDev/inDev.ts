import { serverRuntimeConfig } from "@/config";

import { getFeedBodyCached, getFeedBodyUnCached } from "..";
import { ErrorResponse } from "../publications/types";

import {
	Project,
	FeedPath,
	AllProjects,
	Consultation,
	InConsultationProjects,
	ProjectDetail,
	IndevPanel,
} from "./types";

export * from "./types";

const cacheKeyPrefix = "inDev",
	{ defaultTTL, longTTL } = serverRuntimeConfig.cache,
	{ origin, apiKey } = serverRuntimeConfig.feeds.inDev;

export const getAllProjects = async (): Promise<Project[]> =>
	await getFeedBodyCached<Project[]>(
		cacheKeyPrefix,
		FeedPath.AllProjects,
		defaultTTL,
		async () =>
			(
				await getFeedBodyUnCached<AllProjects>(
					origin,
					FeedPath.AllProjects,
					apiKey
				)
			).embedded.niceIndevIndevelopmentProject
	);

export const getAllConsultations = async (): Promise<Consultation[]> =>
	await getFeedBodyCached<Consultation[]>(
		cacheKeyPrefix,
		FeedPath.InConsultationProjects,
		defaultTTL,
		async () =>
			(
				await getFeedBodyUnCached<InConsultationProjects>(
					origin,
					FeedPath.InConsultationProjects,
					apiKey
				)
			)?.embedded?.niceIndevInconsultationProduct || []
	);

/**
 * Gets a project detail.
 *
 */

export const getProjectDetail = async (
	inDevReference: string
): Promise<ProjectDetail | ErrorResponse> =>
	//TODO don't cache error response
	await getFeedBodyCached<ProjectDetail | ErrorResponse>(
		cacheKeyPrefix,
		FeedPath.ProjectDetail + inDevReference,
		longTTL,
		async () =>
			await getFeedBodyUnCached<ProjectDetail | ErrorResponse>(
				origin,
				FeedPath.ProjectDetail + inDevReference,
				apiKey
			)
	);
