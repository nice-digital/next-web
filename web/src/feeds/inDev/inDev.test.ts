/**
 * @jest-environment node
 */
import { cache } from "@/cache";
import { serverRuntimeConfig } from "@/config";

import { getAllProjects, getAllConsultations } from "./inDev";

const cacheWrapMock = cache.wrap as jest.Mock;

describe("feeds/inDev", () => {
	describe("getAllProjects", () => {
		it("should return value from cache", async () => {
			const data = {};
			cacheWrapMock.mockResolvedValue(data);

			const retVal = await getAllProjects();

			expect(cacheWrapMock).toHaveBeenCalledTimes(1);
			expect(retVal).toBe(data);
		});

		it("should use correct cache key and default TTL", async () => {
			await getAllProjects();

			expect(cacheWrapMock).toHaveBeenCalledWith(
				"next-web:tests:indev:/gidprojects/all",
				expect.any(Function),
				{ ttl: serverRuntimeConfig.cache.defaultTTL }
			);
		});

		it("should load raw feed data with empty cache", async () => {
			await getAllProjects();
			const data: Awaited<ReturnType<typeof getAllProjects>> =
				await cacheWrapMock.mock.calls[0][1]();

			expect(data).toHaveLength(887);
			expect(data[0]).toMatchSnapshot();
		});
	});

	describe("getAllConsultations", () => {
		it("should return value from cache", async () => {
			const data = {};
			cacheWrapMock.mockResolvedValue(data);

			const retVal = await getAllConsultations();

			expect(cacheWrapMock).toHaveBeenCalledTimes(1);
			expect(retVal).toBe(data);
		});

		it("should use correct cache key and default TTL", async () => {
			await getAllConsultations();

			expect(cacheWrapMock).toHaveBeenCalledWith(
				"next-web:tests:indev:/inconsultationprojects",
				expect.any(Function),
				{ ttl: serverRuntimeConfig.cache.defaultTTL }
			);
		});

		it("should load raw feed data with empty cache", async () => {
			await getAllConsultations();
			const data: Awaited<ReturnType<typeof getAllConsultations>> =
				await cacheWrapMock.mock.calls[0][1]();

			expect(data).toHaveLength(20);
			expect(data[0]).toMatchSnapshot();
		});
	});
});
