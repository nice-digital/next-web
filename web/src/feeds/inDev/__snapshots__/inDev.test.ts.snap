// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feeds/inDev getAllConsultations should load raw feed data with empty cache 1`] = `
Object {
  "AllowedRole": "CHTETeam",
  "AreasOfInterestList": Array [],
  "ConsultationId": 1229,
  "ConsultationName": "Draft guidance",
  "ConsultationType": "Draft guidance",
  "DevelopedAs": null,
  "ETag": null,
  "EndDate": "2021-07-16T17:00:00",
  "FirstChapterSlugOfFirstConvertedDocument": null,
  "FirstConvertedDocumentId": null,
  "HasDocumentsWhichAllowConsultationComments": true,
  "IsCHTE": true,
  "OrigProjectReference": null,
  "PartiallyUpdatedProjectReference": null,
  "Process": "MT",
  "ProductTypeName": "Medical technologies guidance",
  "ProjectType": "MTG",
  "Reference": "GID-MT553",
  "RelevantTo": null,
  "ResourceTitleId": "html-content",
  "ShowExpressionOfInterestSubmissionQuestion": false,
  "StartDate": "2021-06-18T00:00:00",
  "Title": "GID-MT553 Synergo for non-muscle-invasive bladder cancer",
  "_links": Object {
    "self": Array [
      Object {
        "href": "/guidance/GID-MT553/consultation/html-content",
      },
    ],
  },
}
`;

exports[`feeds/inDev getAllProjects should load raw feed data with empty cache 1`] = `
Object {
  "AreasOfInterestList": Array [],
  "CreatedDate": "2014-07-16T13:14:34",
  "DevelopedAs": null,
  "ETag": "01000000-0000-001B-0000-000000005402",
  "EvidenceAssessmentGroup": null,
  "FirstGoLiveDate": null,
  "IdNumber": null,
  "LastModifiedDate": "2014-07-16T13:14:34",
  "Process": "MTA",
  "ProductTypeName": "Technology appraisal guidance",
  "ProjectGroup": "Guidance",
  "ProjectType": "TA",
  "PublishedDate": null,
  "Reference": "GID-TAG377",
  "RelevantTo": null,
  "Status": "Discontinued",
  "Title": "Atrial fibrillation - ximelagatran [ID376]",
  "_links": Object {
    "self": Array [
      Object {
        "href": "/project/GID-TAG377",
      },
    ],
  },
}
`;
